Move pkg/forwarder to comp/forwarder/defaultforwarder

ogaca-dd opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?
This PR moves `pkg/forwarder` to `comp/forwarder/defaultforwarder`. 
It also uses the  `forwarder` component in unit tests and removes the usage of `config.Data`. It is a followup of https://github.com/DataDog/datadog-agent/pull/16479.

Here is the list of remaining tasks:
- Remove the global logger from `/com/forwarder`
- Remove `config.Data` introduced by this PR on existing components (dogstatsd in particular)
- Remove globals from `/comp/forwarder`
- Reduce the API surface of `/comp/forwarder` as a lot of functions and structs are exported.

**This is strongly recommend to review this PR commit by commit as lot of code was moved.**
<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Describe how to test/QA your changes

- cluster agent: Check that the default log file location for the `cluster-agent` and `cluster-agentcloudfoundry` is correct.
- security-agent: Check that the default log file location for the `security-agent` is correct.
- system-probe: Check that the default log file location for the `system-probe` is correct.
- windows agent: Check `winutil.LogEventViewer(config.ServiceName, 0x8000000F, DefaultConfPath)` works as expected.
- Agent shared component: Perform sanity checks on the forwarder. 


### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ogaca-dd committed:
Remove InitAndStartAgentDemultiplexerTest

ogaca-dd committed:
Move /pkg/forwarder to /comp/forwarder

ogaca-dd committed:
Replace config.Datadog by config.Component

ogaca-dd committed:
Move paths to its own package cmd/agent/common/path
- Remove dependency cmd/agent/common from com/core/config (to break circle)
- Code for common_windows is split

ogaca-dd committed:
Fix missing import

ogaca-dd committed:
Use config.Component in unit tests of comp/forwarder
- Use config.Datadog for files in /pkg
- Do not inject config in HTTPTransaction because it is serialized
- fxutil.Test can take variatic fx.Option

ogaca-dd committed:
Move config.Datadog.Set("hostname", "XXX") after config.MockModule because config.MockModule resets the config

AlexandreYang reviewed (approved):
lgtm for ndm

ogaca-dd committed:
Fix: system-probe doesnt' use `cmd/agent/common` so revert the change.

ogaca-dd committed:
Add TODO for NewDefaultForwarder

ogaca-dd committed:
Don't use ddconfig.Datadog in pkg.process

truthbk commented on comp/forwarder/defaultforwarder/transaction/transaction.go:
```diff
@@ -201,7 +201,7 @@ type TransactionsSerializer interface {
 // Transaction represents the task to process for a Worker.
 type Transaction interface {
-	Process(ctx context.Context, client *http.Client) error
+	Process(ctx context.Context, config config.Component, client *http.Client) error
```
I wonder if it wouldn't make more sense to make `config.Component` a field in the different `Transaction`-implementing structs? Just, to avoid pushing it down the call stack as a parameter. This might be something for later.

truthbk commented on pkg/collector/corechecks/snmp/snmp_test.go:
```diff
@@ -2043,16 +2053,17 @@ metric_tags:
 	chk := Check{sessionFactory: sessionFactory}
-	aggregator.InitAndStartAgentDemultiplexerTest(demuxOpts(), "")
+	forwarder := fxutil.Test[defaultforwarder.Component](t, defaultforwarder.MockModule, config.MockModule)
+	coreconfig.Datadog.Set("hostname", "test-hostname")
```
It'd be better if we can try to avoid making calls to globals in our tests. I know this is beyond the scope of this PR, and is something that needs to be addressed in multiple places, just mentioning it here for reference.

ogaca-dd replied:
This code was moved and I didn't want to cleanup all the tests as the PR is quite big.

truthbk reviewed (approved):
Couple of comments, but no blockers, just thoughts. Looks ready for 🚀

ogaca-dd committed:
Fix compilation issues

ogaca-dd committed:
Fix missing imports

ogaca-dd committed:
ci-fix:
- Fix windows build
- Fix TestSendConnectionsMessage

ogaca-dd committed:
Merge branch 'main' of github.com:DataDog/datadog-agent into olivierg/move-forwarder-to-comp

ogaca-dd committed:
Fix conflicts

ogaca-dd committed:
Fix Test_Run_simpleCase
Calling mock config must be done before `checkconfig.SetConfdPathAndCleanProfiles()`.

ogaca-dd committed:
Fix snmp tests
Calling mock config must be done before `checkconfig.SetConfdPathAndCleanProfiles()`. I didn't see the issue as the code try to check for /opt/datadog-agent/etc/conf.d/snmp.d/profiles which exists locally

ogaca-dd committed:
Fix windows linters

just-chillin commented on pkg/process/runner/submitter.go:
```diff
@@ -75,22 +75,22 @@ type CheckSubmitter struct {
 
-func NewSubmitter(cfg ddconfig.ConfigReader, hostname string) (*CheckSubmitter, error) {
-	queueBytes := cfg.GetInt("process_config.process_queue_bytes")
+func NewSubmitter(config config.Component, hostname string) (*CheckSubmitter, error) {
```
Nit: I think we don't want to do this. Ideally for us, library code should be able to be used without fx.

just-chillin reviewed (approved):
Looks good. I realize the freeze is tomorrow, so don't let my comments block this PR
